workspaces/mcp-chat/plugins/mcp-chat-backend/src/services/MCPClientServiceImpl.ts

```diff
@@ -0,0 +1,471 @@
+                  error instanceof Error ? error.message : error
+                }. Please ensure Node.js is properly installed with npx available.`,
+              );
+            }
+          } else if (serverConfig.scriptPath) {
+            // Use script path
+            const isPythonScript = serverConfig.scriptPath.endsWith('.py');
+            const isWindows = process.platform === 'win32';
+
+            if (isPythonScript) {
+              command = isWindows ? 'python' : 'python3';
+            } else {
+              command = process.execPath;
+            }
+            args = [serverConfig.scriptPath, ...(serverConfig.args || [])];
+          } else {
+            throw new Error(
+              `Server config for '${serverConfig.name}' must have either scriptPath, npxCommand, or url`,
+            );
+          }
+
+          transport = new StdioClientTransport({
+            command,
+            args,
+            env: {
+              ...process.env, // Inherit current environment
+              ...serverConfig.env, // Add config-specific env vars
+              // Ensure node is in PATH when using npx
+              ...(serverConfig.npxCommand && {
+                PATH: `${path.dirname(process.execPath)}:${
+                  process.env.PATH || ''
+                }`,
+              }),
+            },
+          });
+        }
+
+        // Connect the client with the appropriate transport
+        await client.connect(transport);
+
+        const toolsResult = await client.listTools();
+
+        const serverTools: ServerTool[] = toolsResult.tools.map(tool => ({
+          type: 'function',
+          function: {
+            name: tool.name,
+            description: tool.description ?? '', // Ensure string
+            parameters: tool.inputSchema,
+          },
+          serverId: serverConfig.name, // Track which server this tool belongs to
+        }));
+
+        allTools.push(...serverTools);
+        this.mcpClients.set(serverConfig.name, client);
+
+        // Record successful connection
+        serverResults.push({
+          ...baseServerConfig,
+          status: {
+            valid: isValid,
+            connected: true,
+          },
+        });
+
+        this.logger.info(
+          `MCP Server '${serverConfig.name}' connected via ${
+            serverConfig.type
+          } with tools: ${serverTools.map(t => t.function.name).join(', ')}`,
+        );
+      } catch (error) {
+        const errorMessage =
+          error instanceof Error ? error.message : String(error);
+
+        // Record failed connection
+        serverResults.push({
+          ...baseServerConfig,
+          status: {
+            valid: isValid,
+            connected: false,
+            error: errorMessage,
+          },
+        });
+
+        this.logger.warn(
+          `Failed to connect to MCP server '${serverConfig.name}': ${errorMessage}`,
+        );
+      }
+    }
+
+    this.tools = allTools;
```

Comment:
I think there might be issue that the tools are only updated on the plugin startup and not after that.. What if some MCP server adds or removes some tool during runtime?  It's the same issue with scaffolder at the moment, see this: https://github.com/backstage/backstage/pull/31244

Reply:
You're right that tools are only fetched when the plugin starts up. However, this is actually the expected behavior for MCP servers. In practice, MCP servers define their available tools as part of their initial connection handshake and don't typically add or remove tools while running. The tools are part of the server's static capabilities, so they shouldn't change during runtime.

If someone does need to refresh the tool list (maybe after updating an MCP server or restarting it), they can just restart the Backstage backend or toggle the server off and on in the UI. 

We could add a manual refresh button in the future if this becomes a common need, but for now, the current approach keeps things simple and avoids unnecessary polling overhead.